Production2021 updates

s-good opened the pull request:
Updates mainly to the analysis routines. In particular:
 
- Add in a simpler version of the code in analyse-results.py. Apart from simplifying the code, the main difference is to force selection of QC checks in a particular order e.g. pick a range check, then a climatology check etc. This was implemented following feedback from colleagues.
- Implement two extra features in db_to_df:
    1) When removing levels flagged by a particular test, allow additional levels around the flagged level to be removed also. This handles the situator where a QC operator or automatic QC test makes slightly different decisions on which levels to flag for the same issue.
    2) Allow removal of levels deeper than a rejected level if it is an XBT. This may be done as once an XBT fails, the rest of the profile may be thought of as suspect.
- Fix minor issues due to change in syntax in the latest version of pandas.

BillMills reviewed (changes requested):
Thanks very much for this, @s-good, particularly the level-counting catch in EN_track, and updating our usage of the pandas API.

Asides from the small tactical points I comment on inline, my biggest concern with this section of the code is the complexity of `db_to_df`; I *think* it's OK after a careful reading, but it's doing a lot with no factorization or unit tests, and took me longer than I'd like to admit to think through with the new padding and XBT cutoff features. It's a cost/benefit question whether we worry about slicing this up into simpler, testable functions before our production run which I will defer to you on, but it's one place for potential future improvement.

Beyond this review, I'm going to need to do a few small test runs to try and understand the new `find_roc_ordered` - I can do so this weekend or Monday latest, and will let you know any questions or concerns with that part then.

s-good commented on analyse-results.py:
```diff
@@ -110,7 +110,10 @@ def find_roc(table,
+                          n_to_extract = n_profiles_to_analyse,
+                          pad=2, 
+                          XBTbelow=True,
+                          mark_training=False)
```
So far I've tended to either use the whole database for training or split the data into separate training and validation databases, which means that I've not needed the mark_training functionality. I do think it is useful to keep it, though, so I've added it into find_roc_ordered in the latests commits.

s-good commented on catchall.py:
```diff
@@ -72,7 +72,10 @@ def amend(combo, df):
+                      n_to_extract = samplesize,
+                      pad=2, 
+                      XBTbelow=True,
+                      mark_training=False)
```
I've added this as a command line argument and removed it from db_to_df.

s-good commented on util/dbutils.py:
```diff
@@ -108,28 +112,28 @@ def db_to_df(table,
+        df.columns = ['uid', 'Truth'] + testNamesSave + ['probe']
         df = df.astype({'uid': 'int'})
+        df = df.astype({'probe': 'float'})
         if filter_on_wire_break_test:
```
This was something I added in recently. If a probe is an XBT, the human QC operator may choose to remove the rest of the profile after a level has been flagged. The probe type is now extracted from the database to work out if it is an XBT and this line is to ensure that the probe type number is interpreted as a number rather than a string, although I'm not sure if it is necessary or not. I was just following the way that uid is set to be integer in the line above.

BillMills replied:
Sorry, my comment wasn't clear - I agree with the typecasting change you're making here, I was suggesting we remove `filter_on_wire_break_test` as an unnecessary duplication of the generic test filtering logic - but I see with your recent changes you're calling out to the new factorization of that logic, reducing the duplication. Looks good!

s-good commented on util/dbutils.py:
```diff
@@ -161,18 +181,17 @@ def db_to_df(table,
-                        for j in range(1, len(df.columns)):
+                        for j in range(1, len(df.columns) - 1):
                             # Retain only the levels that passed testname.
                             # Some QC tests may return only one value so check for this.
```
This code has been rewritten and the comment removed.

s-good commented:
Thanks for the review, @BillMills. I've had a go at rewriting db_to_df to make it clearer and have added in testing. One part of it I haven't been able to test is the looping over subsections of the database. I changed this from doing 1000 loops of nprofiles/1000 profiles to doing nprofiles/1000 loops of 1000 profiles. I thought this would be more efficient for smaller databases. I think it is correct, but it is difficult to test if it is working as even the testing on 2 profiles I've added into tests/dbutils_validation.py is quite slow. I've also made the mark_training changes you asked for - I'd like to keep this functionality as I think it is useful to include.

BillMills commented:
@s-good thanks for these updates, looks great! Per your comments, I extended this in #271 with a simple unit test to at least check that `db_to_df` doesn't change based on batch size, but these changes are good to go, feel free to merge.
